File: collect_chat_ideas.py
#!/usr/bin/env python3
"""
Сбор идей по доработке из всех чатов.

Анализирует последние сообщения из чатов и извлекает:
- Запросы на новые функции
- Проблемы пользователей
- Упоминания инструментов
- Идеи по улучшению
"""

import asyncio
import os
from pathlib import Path
from collections import defaultdict
from datetime import datetime, timedelta
from typing import List, Dict, Set
import re

from telethon import TelegramClient
from dotenv import load_dotenv

load_dotenv()

API_ID = int(os.getenv("TELEGRAM_API_ID"))
API_HASH = os.getenv("TELEGRAM_API_HASH")
SESSION_FILE = os.getenv("TELEGRAM_SESSION_FILE", "telegram_session")


# Ключевые слова для поиска идей
FEATURE_KEYWORDS = [
    "надо", "нужно", "хочу", "было бы классно", "предлагаю",
    "идея", "можно добавить", "круто было бы", "не хватает",
    "попросил", "запросил", "автоматизировать", "интеграция",
    "add", "feature", "implement", "need", "want", "idea"
]

PROBLEM_KEYWORDS = [
    "не работает", "сломалось", "баг", "ошибка", "проблема",
    "не могу", "не получается", "почему", "как сделать",
    "broken", "bug", "error", "issue", "problem", "doesn't work"
]

TOOL_KEYWORDS = [
    "cursor", "claude", "gpt", "chatgpt", "copilot", "github",
    "linear", "notion", "slack", "telegram", "n8n", "zapier",
    "api", "webhook", "bot", "automation", "ai", "ml",
    "docker", "kubernetes", "python", "rust", "typescript"
]


class IdeaCollector:
    """Сборщик идей из чатов."""

    def __init__(self):
        self.client: TelegramClient = None
        self.ideas: Dict[str, List[Dict]] = defaultdict(list)
        self.problems: Dict[str, List[Dict]] = defaultdict(list)
        self.tools: Dict[str, Set[str]] = defaultdict(set)

    async def __aenter__(self):
        """Connect to Telegram."""
        self.client = TelegramClient(SESSION_FILE, API_ID, API_HASH)
        await self.client.connect()

        if not await self.client.is_user_authorized():
            raise RuntimeError("Not authorized. Run init_session.py first.")

        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """Disconnect from Telegram."""
        if self.client:
            await self.client.disconnect()

    async def collect_from_chat(self, chat_id: str, days_back: int = 7, limit: int = 500):
        """Collect ideas from a single chat.

        Args:
            chat_id: Chat identifier (@username or ID)
            days_back: How many days back to analyze
            limit: Maximum number of messages
        """
        print(f"\n📊 Анализирую чат: {chat_id}")

        try:
            entity = await self.client.get_entity(chat_id)
            chat_name = getattr(entity, 'title', chat_id)
        except Exception as e:
            print(f"❌ Не могу получить чат {chat_id}: {e}")
            return

        offset_date = datetime.now() - timedelta(days=days_back)

        message_count = 0
        async for message in self.client.iter_messages(
            entity,
            limit=limit,
            offset_date=offset_date
        ):
            if not message.message:
                continue

            message_count += 1
            text = message.message.lower()

            # Поиск идей
            for keyword in FEATURE_KEYWORDS:
                if keyword in text and len(message.message) > 20:
                    self.ideas[chat_name].append({
                        'text': message.message,
                        'date': message.date,
                        'sender': await self._get_sender_name(message),
                        'id': message.id
                    })
                    break

            # Поиск проблем
            for keyword in PROBLEM_KEYWORDS:
                if keyword in text and len(message.message) > 20:
                    self.problems[chat_name].append({
                        'text': message.message,
                        'date': message.date,
                        'sender': await self._get_sender_name(message),
                        'id': message.id
                    })
                    break

            # Поиск упоминаний инструментов
            for tool in TOOL_KEYWORDS:
                if re.search(r'\b' + re.escape(tool) + r'\b', text):
                    self.tools[chat_name].add(tool)

        print(f"  ✅ Проанализировано сообщений: {message_count}")
        print(f"  💡 Найдено идей: {len(self.ideas[chat_name])}")
        print(f"  ⚠️  Найдено проблем: {len(self.problems[chat_name])}")
        print(f"  🔧 Упомянуто инструментов: {len(self.tools[chat_name])}")

    async def _get_sender_name(self, message) -> str:
        """Get sender name from message."""
        if not message.sender:
            return "Unknown"

        sender = message.sender
        if hasattr(sender, 'first_name'):
            return f"{sender.first_name or ''} {sender.last_name or ''}".strip() or "Unknown"
        elif hasattr(sender, 'title'):
            return sender.title
        return "Unknown"

    def generate_report(self, output_file: Path):
        """Generate markdown report with collected ideas.

        Args:
            output_file: Path to save report
        """
        lines = []

        # Header
        lines.append("# 💡 Идеи по доработке из чатов")
        lines.append("")
        lines.append(f"**Дата анализа:** {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        lines.append("")

        # Summary
        total_ideas = sum(len(ideas) for ideas in self.ideas.values())
        total_problems = sum(len(probs) for probs in self.problems.values())
        all_tools = set()
        for tools in self.tools.values():
            all_tools.update(tools)

        lines.append("## 📊 Общая статистика")
        lines.append("")
        lines.append(f"- **Чатов проанализировано:** {len(self.ideas)}")
        lines.append(f"- **Идей найдено:** {total_ideas}")
        lines.append(f"- **Проблем найдено:** {total_problems}")
        lines.append(f"- **Уникальных инструментов:** {len(all_tools)}")
        lines.append("")

        # Ideas by chat
        lines.append("## 💡 Идеи по чатам")
        lines.append("")

        for chat_name, ideas in sorted(self.ideas.items()):
            if not ideas:
                continue

            lines.append(f"### {chat_name}")
            lines.append("")
            lines.append(f"Найдено идей: {len(ideas)}")
            lines.append("")

            # Show top 10 ideas
            for idea in sorted(ideas, key=lambda x: x['date'], reverse=True)[:10]:
                date_str = idea['date'].strftime('%Y-%m-%d %H:%M')
                lines.append(f"**[{date_str}] {idea['sender']}:**")
                lines.append(f"> {idea['text'][:200]}{'...' if len(idea['text']) > 200 else ''}")
                lines.append("")

        # Problems by chat
        lines.append("## ⚠️ Проблемы по чатам")
        lines.append("")

        for chat_name, problems in sorted(self.problems.items()):
            if not problems:
                continue

            lines.append(f"### {chat_name}")
            lines.append("")
            lines.append(f"Найдено проблем: {len(problems)}")
            lines.append("")

            # Show top 10 problems
            for problem in sorted(problems, key=lambda x: x['date'], reverse=True)[:10]:
                date_str = problem['date'].strftime('%Y-%m-%d %H:%M')
                lines.append(f"**[{date_str}] {problem['sender']}:**")
                lines.append(f"> {problem['text'][:200]}{'...' if len(problem['text']) > 200 else ''}")
                lines.append("")

        # Tool mentions
        lines.append("## 🔧 Упоминаемые инструменты")
        lines.append("")

        for chat_name, tools in sorted(self.tools.items()):
            if not tools:
                continue

            lines.append(f"### {chat_name}")
            lines.append("")
            lines.append(f"Инструменты: {', '.join(sorted(tools))}")
            lines.append("")

        # Popular tools across all chats
        lines.append("## 🌟 Топ инструментов по популярности")
        lines.append("")

        tool_counts = defaultdict(int)
        for tools in self.tools.values():
            for tool in tools:
                tool_counts[tool] += 1

        for tool, count in sorted(tool_counts.items(), key=lambda x: x[1], reverse=True)[:20]:
            lines.append(f"- **{tool}**: упомянут в {count} чатах")

        lines.append("")

        # Action items
        lines.append("## 🎯 Рекомендации по доработке")
        lines.append("")
        lines.append("На основе анализа чатов, рекомендую:")
        lines.append("")
        lines.append("1. **Приоритетные фичи:** Проанализировать самые частые запросы")
        lines.append("2. **Критичные проблемы:** Исправить упомянутые баги")
        lines.append("3. **Интеграции:** Рассмотреть интеграции с популярными инструментами")
        lines.append("4. **Документация:** Создать гайды по решению частых проблем")
        lines.append("")

        # Save
        output_file.parent.mkdir(parents=True, exist_ok=True)
        output_file.write_text('\n'.join(lines), encoding='utf-8')

        print(f"\n✅ Отчёт сохранён: {output_file}")


async def main():
    """Main function."""
    import argparse

    parser = argparse.ArgumentParser(description="Собрать идеи по доработке из чатов")
    parser.add_argument(
        "--chats",
        nargs="+",
        default=["@vibecod3rs"],
        help="Чаты для анализа (например: @vibecod3rs @channel2)"
    )
    parser.add_argument(
        "--days",
        type=int,
        default=7,
        help="Сколько дней назад анализировать (по умолчанию: 7)"
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=500,
        help="Максимум сообщений на чат (по умолчанию: 500)"
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("analysis_results/chat_ideas.md"),
        help="Путь для сохранения отчёта"
    )

    args = parser.parse_args()

    print("🚀 Начинаю сбор идей из чатов...")
    print(f"📅 Период анализа: последние {args.days} дней")
    print(f"📝 Лимит сообщений: {args.limit} на чат")
    print()

    async with IdeaCollector() as collector:
        for chat in args.chats:
            await collector.collect_from_chat(chat, args.days, args.limit)

        collector.generate_report(args.output)

    print("\n✨ Готово!")


if __name__ == "__main__":
    asyncio.run(main())

File: src/bin/collect_chat_ideas.rs
//! Collect ideas for improvement from all chats.
//!
//! Analyzes recent messages from chats and extracts:
//! - Requests for new features
//! - User problems
//! - Tool mentions
//! - Improvement ideas
//!
//! Usage: cargo run --bin collect_chat_ideas -- [--chats @vibecod3rs] [--days 7] [--limit 500] [--output analysis_results/chat_ideas.md]

use anyhow::Result;
use chrono::{DateTime, Duration, Utc};
use clap::Parser;
use dotenvy::dotenv;
use grammers_client::types::{Message, Peer};
use regex::Regex;
use std::collections::{HashMap, HashSet};
use std::fs;
use std::path::PathBuf;
use telegram_reader::chat::{find_chat, peer_name};
use telegram_reader::{get_client, SessionLock};

#[derive(Parser)]
struct Args {
    /// Chats to analyze (e.g. @vibecod3rs)
    #[arg(long, default_values = &["@vibecod3rs"])]
    chats: Vec<String>,

    /// How many days back to analyze
    #[arg(long, default_value = "7")]
    days: i64,

    /// Maximum number of messages per chat
    #[arg(long, default_value = "500")]
    limit: usize,

    /// Path for saving the report
    #[arg(long, default_value = "analysis_results/chat_ideas.md")]
    output: PathBuf,
}

const FEATURE_KEYWORDS: &[&str] = &[
    "надо", "нужно", "хочу", "было бы классно", "предлагаю", "идея", "можно добавить",
    "круто было бы", "не хватает", "попросил", "запросил", "автоматизировать",
    "интеграция", "add", "feature", "implement", "need", "want", "idea",
];

const PROBLEM_KEYWORDS: &[&str] = &[
    "не работает", "сломалось", "баг", "ошибка", "проблема", "не могу", "не получается",
    "почему", "как сделать", "broken", "bug", "error", "issue", "problem", "doesn't work",
];

const TOOL_KEYWORDS: &[&str] = &[
    "cursor", "claude", "gpt", "chatgpt", "copilot", "github", "linear", "notion",
    "slack", "telegram", "n8n", "zapier", "api", "webhook", "bot", "automation",
    "ai", "ml", "docker", "kubernetes", "python", "rust", "typescript",
];

#[derive(Debug)]
struct Entry {
    text: String,
    date: DateTime<Utc>,
    sender: String,
    id: i32,
}

struct IdeaCollector {
    ideas: HashMap<String, Vec<Entry>>,
    problems: HashMap<String, Vec<Entry>>,
    tools: HashMap<String, HashSet<String>>,
}

impl IdeaCollector {
    fn new() -> Self {
        Self {
            ideas: HashMap::new(),
            problems: HashMap::new(),
            tools: HashMap::new(),
        }
    }

    async fn collect_from_chat(&mut self, client: &grammers_client::Client, chat_id: &str, days_back: i64, limit: usize) -> Result<()> {
        println!("\n📊 Анализирую чат: {}", chat_id);

        let peer = match find_chat(client, chat_id).await {
            Ok(p) => p,
            Err(e) => {
                eprintln!("❌ Не могу найти чат {}: {}", chat_id, e);
                return Ok(());
            }
        };

        let chat_name = peer_name(&peer);
        let offset_date = Utc::now() - Duration::days(days_back);

        let mut messages = client.iter_messages(&peer).limit(limit);
        let mut message_count = 0;

        while let Some(message) = messages.next().await? {
            if message.date() < offset_date {
                break;
            }

            let text = message.text();
            if text.is_empty() {
                continue;
            }

            message_count += 1;
            let text_lower = text.to_lowercase();

            // Check for ideas
            for keyword in FEATURE_KEYWORDS {
                if text_lower.contains(keyword) && text.len() > 20 {
                    let sender = self.get_sender_name(&message).await;
                    self.ideas.entry(chat_name.clone()).or_default().push(Entry {
                        text: text.to_string(),
                        date: message.date(),
                        sender,
                        id: message.id(),
                    });
                    break;
                }
            }

            // Check for problems
            for keyword in PROBLEM_KEYWORDS {
                if text_lower.contains(keyword) && text.len() > 20 {
                    let sender = self.get_sender_name(&message).await;
                    self.problems.entry(chat_name.clone()).or_default().push(Entry {
                        text: text.to_string(),
                        date: message.date(),
                        sender,
                        id: message.id(),
                    });
                    break;
                }
            }

            // Check for tools
            for tool in TOOL_KEYWORDS {
                let re = Regex::new(&format!(r"\b{}\b", regex::escape(tool)))?;
                if re.is_match(&text_lower) {
                    self.tools.entry(chat_name.clone()).or_default().insert(tool.to_string());
                }
            }
        }

        println!("  ✅ Проанализировано сообщений: {}", message_count);
        println!("  💡 Найдено идей: {}", self.ideas.get(&chat_name).map_or(0, |v| v.len()));
        println!("  ⚠️  Найдено проблем: {}", self.problems.get(&chat_name).map_or(0, |v| v.len()));
        println!("  🔧 Упомянуто инструментов: {}", self.tools.get(&chat_name).map_or(0, |v| v.len()));

        Ok(())
    }

    async fn get_sender_name(&self, message: &Message) -> String {
        if let Some(sender) = message.sender() {
            peer_name(&sender)
        } else {
            "Unknown".to_string()
        }
    }

    fn generate_report(&self, output_file: &PathBuf) -> Result<()> {
        let mut lines = Vec::new();

        lines.push("# 💡 Идеи по доработке из чатов".to_string());
        lines.push("".to_string());
        lines.push(format!("**Дата анализа:** {}", Utc::now().format("%Y-%m-%d %H:%M:%S")));
        lines.push("".to_string());

        let total_ideas: usize = self.ideas.values().map(|v| v.len()).sum();
        let total_problems: usize = self.problems.values().map(|v| v.len()).sum();
        let mut all_tools = HashSet::new();
        for t in self.tools.values() {
            for tool in t {
                all_tools.insert(tool);
            }
        }

        lines.push("## 📊 Общая статистика".to_string());
        lines.push("".to_string());
        lines.push(format!("- **Чатов проанализировано:** {}", self.ideas.len()));
        lines.push(format!("- **Идей найдено:** {}", total_ideas));
        lines.push(format!("- **Проблем найдено:** {}", total_problems));
        lines.push(format!("- **Уникальных инструментов:** {}", all_tools.len()));
        lines.push("".to_string());

        lines.push("## 💡 Идеи по чатам".to_string());
        lines.push("".to_string());

        let mut chat_names: Vec<_> = self.ideas.keys().collect();
        chat_names.sort();

        for chat_name in chat_names {
            if let Some(ideas) = self.ideas.get(chat_name) {
                if ideas.is_empty() { continue; }
                lines.push(format!("### {}", chat_name));
                lines.push("".to_string());
                lines.push(format!("Найдено идей: {}", ideas.len()));
                lines.push("".to_string());

                let mut sorted_ideas = ideas.clone();
                sorted_ideas.sort_by(|a, b| b.date.cmp(&a.date));

                for idea in sorted_ideas.iter().take(10) {
                    lines.push(format!("**[{}] {}:**", idea.date.format("%Y-%m-%d %H:%M"), idea.sender));
                    let preview = if idea.text.len() > 200 {
                        format!("{}...", &idea.text[..200])
                    } else {
                        idea.text.clone()
                    };
                    lines.push(format!("> {}", preview.replace('\n', "\n> ")));
                    lines.push("".to_string());
                }
            }
        }

        lines.push("## ⚠️ Проблемы по чатам".to_string());
        lines.push("".to_string());

        let mut prob_chat_names: Vec<_> = self.problems.keys().collect();
        prob_chat_names.sort();

        for chat_name in prob_chat_names {
            if let Some(problems) = self.problems.get(chat_name) {
                if problems.is_empty() { continue; }
                lines.push(format!("### {}", chat_name));
                lines.push("".to_string());
                lines.push(format!("Найдено проблем: {}", problems.len()));
                lines.push("".to_string());

                let mut sorted_problems = problems.clone();
                sorted_problems.sort_by(|a, b| b.date.cmp(&a.date));

                for problem in sorted_problems.iter().take(10) {
                    lines.push(format!("**[{}] {}:**", problem.date.format("%Y-%m-%d %H:%M"), problem.sender));
                    let preview = if problem.text.len() > 200 {
                        format!("{}...", &problem.text[..200])
                    } else {
                        problem.text.clone()
                    };
                    lines.push(format!("> {}", preview.replace('\n', "\n> ")));
                    lines.push("".to_string());
                }
            }
        }

        lines.push("## 🔧 Упоминаемые инструменты".to_string());
        lines.push("".to_string());

        let mut tool_chat_names: Vec<_> = self.tools.keys().collect();
        tool_chat_names.sort();

        for chat_name in tool_chat_names {
            if let Some(tools) = self.tools.get(chat_name) {
                if tools.is_empty() { continue; }
                let mut sorted_tools: Vec<_> = tools.iter().collect();
                sorted_tools.sort();
                let tools_str: Vec<_> = sorted_tools.into_iter().cloned().collect();
                lines.push(format!("### {}", chat_name));
                lines.push("".to_string());
                lines.push(format!("Инструменты: {}", tools_str.join(", ")));
                lines.push("".to_string());
            }
        }

        lines.push("## 🌟 Топ инструментов по популярности".to_string());
        lines.push("".to_string());

        let mut tool_counts = HashMap::new();
        for tools in self.tools.values() {
            for tool in tools {
                *tool_counts.entry(tool).or_insert(0) += 1;
            }
        }

        let mut sorted_counts: Vec<_> = tool_counts.into_iter().collect();
        sorted_counts.sort_by(|a, b| b.1.cmp(&a.1));

        for (tool, count) in sorted_counts.iter().take(20) {
            lines.push(format!("- **{}**: упомянут в {} чатах", tool, count));
        }

        lines.push("".to_string());
        lines.push("## 🎯 Рекомендации по доработке".to_string());
        lines.push("".to_string());
        lines.push("На основе анализа чатов, рекомендую:".to_string());
        lines.push("".to_string());
        lines.push("1. **Приоритетные фичи:** Проанализировать самые частые запросы".to_string());
        lines.push("2. **Критичные проблемы:** Исправить упомянутые баги".to_string());
        lines.push("3. **Интеграции:** Рассмотреть интеграции с популярными инструментами".to_string());
        lines.push("4. **Документация:** Создать гайды по решению частых проблем".to_string());
        lines.push("".to_string());

        if let Some(parent) = output_file.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(output_file, lines.join("\n"))?;

        println!("\n✅ Отчёт сохранён: {:?}", output_file);
        Ok(())
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv().ok();
    tracing_subscriber::fmt::init();

    let args = Args::parse();

    println!("🚀 Начинаю сбор идей из чатов...");
    println!("📅 Период анализа: последние {} дней", args.days);
    println!("📝 Лимит сообщений: {} на чат", args.limit);
    println!();

    let _lock = SessionLock::acquire()?;
    let client = get_client().await?;

    let mut collector = IdeaCollector::new();

    for chat in &args.chats {
        collector.collect_from_chat(&client, chat, args.days, args.limit).await?;
    }

    collector.generate_report(&args.output)?;

    println!("\n✨ Готово!");
    Ok(())
}
